### PR TITLE
Fix negative quantity validation in the admin order editor

### LINKED ISSUE #31068: Adding items to order in backend with negative qty results in error
- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
When editing an order in the backend, it is possible to add a product with a negative quantity. However, after that, the **Update** button does not work anymore.  Error in console: "_An invalid form control with name 'order_item_qty[2767] is not focusable._"

**Expected behavior**
Either it should be made impossible to add a product with a negative quantity, **OR** the Update button should work again after that.

**Actual behavior**
I believe it should not be possible to add products with a negative quantity to begin with.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. In the backend, open an order that can be edited (e.g. "Pending Payment")
2. Add an item to the order with quantity -1 (you have to *type* it, as the arrows will only let you go to 0).
3. Select the Update button.

**Screenshots**
https://d.pr/i/4zDXYS

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by ralucaStan**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

In the admin order editor, typing a negative line item quantity fails silently: the hidden `min="0"` input cancels **Update** with only a console error, and the items panel **Save** persists the value with no server-side check. This PR enforces the quantity minimum in the editor: the browser validation message shows in all flows (Update, panel Save, Add products modal), and the order item AJAX endpoints reject quantities below the minimum via a new internal `ItemQuantityLimits` class.

Closes #31068.

**Existing negative orders must keep working**

Negative line items are valid data for integrations; see [this comment](https://github.com/woocommerce/woocommerce/issues/31068#issuecomment-960313222) (ERP/RMA flows that only need **Update** to work on such orders) and [this one](https://github.com/woocommerce/woocommerce/issues/31068#issuecomment-2152242898) (reconciliation orders created by hand). Blocking those orders would be a regression, so for items that already store a negative quantity the input minimum is floored at the stored value; the order stays editable and only going lower is blocked. Negative quantities are prevented as *new* input going forward. Shops that want to keep entering them can lower the minimum via the `woocommerce_quantity_input_min_admin` filter.

**BC impact**

- `WC_AJAX::add_order_item`, `save_order_items`, and the save passthroughs in `add_order_item`/`remove_order_item` now reject quantities below the admin minimum with a JSON error; this is the bug fix itself. The REST API, `wc_save_order_items()`, and `WC_Abstract_Order::add_product()` are unchanged, so programmatic integrations are unaffected.
- The `woocommerce_quantity_input_min_admin` filter changes are additive: a new `'add'` context, it runs during the AJAX validation now, and its `'edit'` default is the stored quantity when that is already negative (previously always `'0'`).
- Non-numeric returns from `woocommerce_quantity_input_min_admin` no longer switch the minimum off. In the `'edit'` and `'add'` contexts a non-numeric return now falls back to the numeric default. Returning `''` used to render `min=""`, which browsers ignore, so it worked as a way to remove the minimum; that came from the markup rather than from the filter, which documents its value as the minimum amount to use. Extensions relying on it should return a numeric floor instead, for example `'-9999'`; that relaxes both the rendered attribute and the new server-side check, so negative quantities stay enterable end to end. The `'refund'` context is unchanged and still passes the filtered value straight to the markup.
- New JS event `wc_backbone_modal_before_response` in `backbone-modal.js`: an additive cancelable step in `addButton()` — listeners call `event.preventDefault()` to keep the modal open, which is how the modal validation hooks in; other modals are unaffected since an uncanceled event keeps them passing.
- Client-side, only the quantity **minimum** is enforced (`validity.rangeUnderflow`); `step` and `max` keep their previous behaviour.
- The add products modal input keeps its hardcoded `min="0"`: no product is selected when the row renders, and the filter documents `$product` as `WC_Product`, so it is not applied with `false` there. Minimums raised via the filter's `'add'` context are enforced server-side on the actual add, and the rejection surfaces through the metabox's existing `window.alert` error pattern, same as the other order-items endpoints. Showing server-side errors inline would mean keeping the modal open across the AJAX request — a follow-up, not this PR.
- Out of scope by decision: the Recalculate flow (`woocommerce_calc_line_taxes`) is not guarded; it is unreachable with unsaved quantity edits through the normal UI, so only direct requests can route a below-minimum quantity through it. Follow-up candidate.

**A note on the approach**

I'm new to this area of the codebase, and I built this with Claude Code, setting the approach myself. I took care with testing and kept the changes scoped, but I'd appreciate the reviewer's input on the approach itself, not only the diff; do let me know if the floor or the filter contract should work differently.

### Screenshots or screen recordings:

The visible change is the browser-native "Value must be greater than or equal to 0" message on the quantity inputs; recording to be added.

### How to test the changes in this Pull Request:

**Existing negative orders (must keep working):**

1. Create one: `wp wc shop_order create --user=admin --status=pending --line_items='[{"product_id":<ID>,"quantity":-3}]'` and open it in the editor. The quantity input renders `min="-3"`.
2. Click **Update** without touching anything: the order saves normally. On trunk the same click does nothing and logs `An invalid form control ... is not focusable`; this PR un-bricks those orders.
3. Pencil-edit the line: leaving `-3` or raising it (`-2`, `0`, positive) saves; `-4` is blocked with the validation message.

**Adding new product quantities:**

4. **Add item(s) → Add product(s)**, pick a product, type `-3`, click **Add**: the modal stays open with the validation message. Pressing **Enter** instead behaves the same.
5. Change to `2` and click **Add**: the product is added normally. A direct `woocommerce_add_order_item` POST with a negative `qty` gets the same JSON error.

**The save paths:**

6. On a normal order, pencil-edit a line, type `-1`, click the panel **Save**: message on the input, no request sent, quantity unchanged after reload.
7. With `-1` still in the field, click **Update**: the field stays visible and focused with the message; the order is not saved.
8. Delete passthrough: pencil-edit line A to `-1`, trash line B, confirm. Deleting saves the in-progress edits first, so the request is rejected and line B is not deleted.
9. Bypass the client: POST to `admin-ajax.php` with `action=woocommerce_save_order_items` and `order_item_qty[<item_id>]=-1`. The response is `{"success":false,"data":{"error":"The quantity of \"...\" must be 0 or higher."}}` (real quotes, no HTML entities) and nothing persists.
10. Regression: normal edits, refunds, and the filter keep working; with an mu-plugin returning `'-9999'` from `woocommerce_quantity_input_min_admin`, `-5` is accepted end-to-end.

### Testing that has already taken place:

- 17 PHP tests pass in wp-env: `ItemQuantityLimitsTest` (12, covering the floor, the filter contract, decimal stock, and the handling of foreign and fee item ids) and 5 new `WC_AJAX_Test` cases.
- 3 Playwright e2e tests pass locally (`tests/e2e/tests/order/order-item-quantity-validation.spec.ts`).
- The AJAX matrix was verified with direct `admin-ajax.php` requests on a local Studio site (HPOS), including the floor cases.
- `lint:changes:branch` and PHPStan are clean; the PHPStan baseline shrank by three entries.

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

Created manually.

</details>

### Release Communication

Select the release communication labels this PR needs:

- [ ] **Feature Highlight** - For user-facing features (what changed, user impact)
- [x] **Developer Advisory** - For developer-facing changes (what changed, how to detect, actions needed)

<details>
<summary>When to use each?</summary>

**Feature Highlight**: New features, UI changes, or improvements that merchants/store owners will notice.

- Example: "New bulk editing for products", "Improved checkout performance"

**Developer Advisory**: Breaking changes, deprecations, or changes that affect themes/plugins/extensions.

- Example: "Hook signature change", "Deprecated filter", "REST API field removed"

Selected labels are applied when the pull request is merged.
</details>

### Use of AI Tools

I built this with Claude Code from a plan I prepared, over several review and revision rounds; I reviewed every diff and ran the tests myself. I'm new to this area, so reviewer scrutiny on the approach is welcome.

### CODE REVIEW

File: plugins/woocommerce/client/legacy/js/admin/backbone-modal.js [hunk 1]
```diff
@@ -109,6 +109,16 @@
...
 		addButton: function( e ) {
+			var validation = { valid: true };
+
+			// Allow listeners to veto the response, e.g. to validate inputs
+			// and keep the modal open. Covers click, touch and keyboard paths.
+			$( document.body ).trigger( 'wc_backbone_modal_before_response', [ this._target, this.$el, validation ] );
+
+			if ( ! validation.valid ) {
+				return;
+			}
```
Comment (opr) [inline plugins/woocommerce/client/legacy/js/admin/backbone-modal.js]: The pre-response hook makes sense because the Backbone modal fires the response and closes unconditionally otherwise, but maybe it should be a cancelable jQuery event, as opposed to passing a mutable `{ valid: true }` object.

Listeners could call `event.preventDefault()`, and `addButton()` could check `event.isDefaultPrevented()`, making cancellation more idiomatic, explicit, and impossible for another listener to accidentally undo.

**Review by opr — CHANGES_REQUESTED** [review]
Hi, I did some testing and the happy path plus a couple of edge cases:
- Setting to 0 quantity
- Starting with a negative quantity, then going positive, then trying to go back to negative

and it looks to be working OK, however CodeRabbit does have some valid points so I would suggest fixing these then I'll re-review.

File: plugins/woocommerce/client/legacy/js/admin/backbone-modal.js (hunk 1 above)
Comment (ralucaStan) [reply plugins/woocommerce/client/legacy/js/admin/backbone-modal.js]: fixed 2e90cffcfc0e27a3b76c823814d272fe9bb85652

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-order.js [hunk 2]
```diff
@@ -1212,6 +1254,25 @@ jQuery( function ( $ ) {
...
+				}
+
+				var qtyInputBelowMin = wc_meta_boxes_order_items.find_input_with_qty_below_min(
+					$modal[ 0 ].querySelectorAll( 'input[name="item_qty"]' )
+				);
+
+				if ( qtyInputBelowMin ) {
+					qtyInputBelowMin.reportValidity();
+					validation.valid = false;
+				}
```
[the remark this answers is not in the document]
Comment (ralucaStan) [reply plugins/woocommerce/client/legacy/js/admin/meta-boxes-order.js:1273]: 1efe9f39821e6b5f855eb75eca24e015124288bb fixed this by aligning the quantities on the front-end. this is the minimal change to fix the bug without editing the existing mechanisma of the submit + close popup on add click.

File: plugins/woocommerce/includes/admin/meta-boxes/views/html-order-items.php [hunk 3]
```diff
@@ -436,9 +436,20 @@
...
+								/**
+								 * This filter is documented in includes/admin/meta-boxes/views/html-order-item.php
+								 *
+								 * No product is selected yet when the modal row is rendered, so the
+								 * product argument is false.
+								 *
+								 * @since 5.8.0
+								 */
+								$add_item_min = apply_filters( 'woocommerce_quantity_input_min_admin', '0', false, 'add' );
+								$add_item_min = is_numeric( $add_item_min ) ? (string) $add_item_min : '0';
```
[the remark this answers is not in the document]
Comment (ralucaStan) [reply plugins/woocommerce/includes/admin/meta-boxes/views/html-order-items.php]: Because we don't know the product in the UI, it's user selected, I've decied to not inforce the filter at the UI level in the modal. 

The validation still happens server-side, and the merchant will see a notice that the minimum must be respected.

The UX is not ideal, but this is legacy code, and improving the experience is beyond the scope here.

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-order.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ralucaStan) [reply plugins/woocommerce/client/legacy/js/admin/meta-boxes-order.js:1273]: We're not using the min filter on the input anymore, so the bug no longer happens. see https://github.com/woocommerce/woocommerce/pull/67921/changes#r3879361428 for reasons.

File: plugins/woocommerce/tests/php/src/Internal/Admin/Orders/ItemQuantityLimitsTest.php [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+		$order = WC_Helper_Order::create_order();
+		$items = array_values( $order->get_items() );
+
+		$this->expectException( \Exception::class );
+		$this->sut->validate_posted_item_quantities(
+			$order,
+			array(
+				'order_item_qty' => array( $items[0]->get_id() => '-0.5' ),
+			)
+		);
```
[the remark this answers is not in the document]
Comment (ralucaStan) [reply plugins/woocommerce/tests/php/src/Internal/Admin/Orders/ItemQuantityLimitsTest.php:220]: No leak here — WP_UnitTestCase backs up $wp_filter in set_up() and _restore_hooks() rebuilds it in tear_down(), which runs even when the test ends on the expected exception. The swap is reverted by the framework after every test, so the try/finally would duplicate that.

File: plugins/woocommerce/client/legacy/js/admin/backbone-modal.js [hunk 5]
```diff
@@ -109,6 +109,16 @@
...
+			// Allow listeners to cancel the response via event.preventDefault(),
+			// e.g. to validate inputs and keep the modal open. Covers click,
+			// touch and keyboard paths.
+			var beforeResponse = $.Event( 'wc_backbone_modal_before_response' );
+			$( document.body ).trigger( beforeResponse, [ this._target, this.$el ] );
+
+			if ( beforeResponse.isDefaultPrevented() ) {
+				return;
+			}
+
```
[the remark this answers is not in the document]
Comment (ralucaStan) [reply plugins/woocommerce/client/legacy/js/admin/backbone-modal.js:121]: wc_backbone_modal_before_response is introduced by this PR

**Comment by ralucaStan** [comment]
> and it looks to be working OK, however CodeRabbit does have some valid points so I would suggest fixing these then I'll re-review.

@opr Thank you for the review. I've fixed the 🐰 comments. Unfortunately, it's too complicated to enforce the filter minimum in the modal UI without massive changes, which are beyond the scope of this PR. 

The minimum is still enforced by the server, and the merchant sees errors, just as before, it's a trunk behaviour.
Screenshot 2026-08-28 at 11 20 40

**Review by opr — APPROVED** [review]
@ralucaStan thanks for addressing the comments, I tested the event interception and it works. Understandable about it being too large of a change to enforce minima client-side in the modal, so I'm happy with skipping that.
